h/settings.py

```diff
@@ -13,54 +14,76 @@ class SettingError(Exception):
-
-
-class EnvSetting(object):
-
-    """An (optionally typed) setting read from an environment variable."""
-
-    def __init__(self, setting, varname, type=None):
-        self.setting = setting
-        self.varname = varname
-        if type is not None:
-            self.type = type
-        else:
-            self.type = str
-
-    def __call__(self, environ):
-        if self.varname in environ:
+class SettingsManager(object):
+    def __init__(self, settings=None, environ=None):
+        self.settings = settings or {}
+        self._environ = environ or os.environ
+
+    def set(self,  # pylint: disable=too-many-arguments
+            name,
+            envvar,
+            type_=str,
+            required=False,
+            default=None,
+            deprecated_msg=None):
+        """
+        Set settings[name] to envvar or default.
+
+        Environment variables override config file settings and
+        defaults: if the named envvar is in the environment then set
+        settings[name] to the value of the envvar.
+
+        Config file settings override defauits: if the envvar isn't
+        in the environment, and the named setting is already present
+        in settings, then leave settings[name] unchanged.
+
+        Defaults are used if the setting isn't in the environment or
+        the config file: if envvar isn't in the environment and name
+        isn't in settings then set settings[name] to the given default
+        value.
+
+        If envvar isn't in the environment and name isn't in settings
+        and no default is given then leave settings[name] unset. If
+        this happens and required=True then raise SettingError.
+
+        If deprecation_msg is not None and the envvar is in the
+        environment then log a warning envvar deprecation message.
+
+        :param name: the name of the pyramid config setting
+        :type name: str
+        :param envvar: the environment variable name
+        :type envvar: str
+        :param type_: the type to type cast the envvar or default value
+        :param required: True if the the pyramid config setting is required
+        :type required: bool
+        :param default: a default value to use if the envvar isn't set
+        :param deprecated_msg: a deprecated envvar setting message to display
+        :type deprecated_msg: str
+        :raises SettingsError: if required and not set or type casting fails
+        """
+        val = None
+        cast_message = None
+        if envvar in self._environ:
+            if deprecated_msg:
+                log.warn('use of envvar %s is deprecated: %s',
+                         envvar,
+                         deprecated_msg)
+            val = self._environ[envvar]
+            cast_message = "environment variable {}={}".format(envvar, val)
+        elif default and name not in self.settings:
+            val = default
+            cast_message = "{}'s default {}".format(name, val)
```

Comment:
After fixing `type_` in `config.py` and running `make dev` again I'm getting a `UnicodeDecodeError` here. At least some of the settings (for example `DEFAULT_SALT`) are byte strings rather than unicode strings, but your error message here is unicode, so the string formatting crashes.